vowpalwabbit/slates_label.cc

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "slates_label.h"
+
+#include "cache.h"
+#include "parser.h"
+#include "vw_string_view.h"
+#include "constant.h"
+
+#include <numeric>
+
+namespace slates
+{
+void default_label(void* v);
+
+#define READ_CACHED_VALUE(DEST, TYPE)                            \
+  next_read_size = sizeof(TYPE);                                 \
+  if (cache.buf_read(read_ptr, next_read_size) < next_read_size) \
+    return 0;                                                    \
+  DEST = *(TYPE*)read_ptr;                                       \
+  read_count += sizeof(TYPE);
+
+#define WRITE_CACHED_VALUE(VALUE, TYPE)           \
+  *(TYPE*)c = VALUE; \
```

Comment:
I worry a bit these macros require specific stack state that is configured outside of them (and uses slightly different names for things: `c` vs `read_ptr`). Consider creating a pair of macros that can be used to set it up, e.g. `READ_CACHE(cache) / WRITE_CACHE(cache, size)`.

Reply:
Yeah I understand that. But given they are defined in a cc file, I am not worried about them leaking and they are just used to not repeat myself (and cause bugs, like I did when I first did this -_-)